ipv8/messaging/interfaces/endpoint.py

```diff
@@ -28,19 +32,44 @@ def add_listener(self, listener):
         """
         if not isinstance(listener, EndpointListener):
             raise IllegalEndpointListenerError(listener)
-        self._listeners.append(listener)
+        with self.listener_update_lock:
+            self._listeners.append(listener)
+            for prefix in self._prefix_map:
+                self._prefix_map[prefix].append(listener)
```

Comment:
Why add the listener to both `self._listeners` and to `self._prefix_map`?

Reply:
Speed upon receiving a packet.

It takes longer to add and remove, but when a packet is received it instantly has all the listeners it needs to deliver the packet to.

Reply:
Oh, and this will use more memory.

Reply:
Basically, we sacrifice memory and complexity for a marginal speed improvement.